presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java

```diff
@@ -17,24 +17,28 @@
 
 import javax.annotation.concurrent.ThreadSafe;
 
-import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.LongConsumer;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-class PageReference
+final class PageReference
 {
+    private static final AtomicIntegerFieldUpdater<PageReference> REFERENCE_COUNT_UPDATER = AtomicIntegerFieldUpdater.newUpdater(PageReference.class, "referenceCount");
+
     private final Page page;
-    private final Runnable onFree;
-    private final AtomicInteger referenceCount;
+    private final LongConsumer onPageReleased;
+    private volatile int referenceCount;
 
-    public PageReference(Page page, int referenceCount, Runnable onFree)
+    public PageReference(Page page, int referenceCount, LongConsumer onPageReleased)
```

Comment:
Changing from `Runnable` to `LongConsumer` doesn't improve anything. Undo. Alternatively make it `Consumer<Page>` as it's more explicit than generic long.

Reply:
`Runnable` requires that each `PageReference` instance is initialized with a lambda that captures state from the creating context in order to perform cleanup. a single LongConsumer can be created per exchange sink and used for all created instances.

`Consumer<Page>` might be ok, but also opens the door to accidental ambiguity like whether `Page#getRetainedSizeInBytes()` or `Page#getSizeInBytes()` should be used as the decrement.

Reply:
> Runnable requires that each PageReference instance is initialized with a lambda that captures state from the creating context in order to perform cleanup. a single LongConsumer can be created per exchange sink and used for all created instances.

You could make it `Consumer<PageReference>`
Then lambda can be:
```
reference -> doSomething(reference.getRetainedSizeInBytes()
```
It's cleaner and more explicit than generic long.